### PR TITLE
feat(reader): Add panel-by-panel navigation for comics and manga

### PR DESCRIPTION
**Pull request by vmanojkumar**
Implements intelligent panel detection and navigation for scanned comics, manga, and other documents with panel-based layouts.

Features:
- Automatic panel detection using K2pdfopt connected component analysis
- 8 reading direction presets (LTR/RTL x Top-to-Bottom/Bottom-to-Top)
- Smart panel sorting with 30% vertical overlap threshold for row detection
- Panel filtering to exclude small artifacts and page-spanning elements
- Adjacent panel merging to combine split panels
- Touch gesture navigation (tap left/right screen halves)
- Debug visualization mode for panel boundary troubleshooting
- Configurable settings preserved per-document

Integration:
- Accessible via Reading menu > Panel Navigation
- Per-document settings stored alongside other reader settings

Testing:
- 46 unit tests covering sorting, filtering, merging, and edge cases
- Tested on Kindle Paperwhite 3 (7th Gen)

Documentation:
- New doc/PanelNavigation.md with complete feature guide
- Updated Hacking.md and Unit_tests.md

Depends on: koreader/koreader-base PR (getAllPanelsFromPage in koptcontext)
https://github.com/koreader/koreader-base/pull/2257

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/14860)

### CODE REVIEW

File: doc/PanelNavigation.md [hunk 1]
```diff
@@ -0,0 +1,192 @@
+# Panel Navigation
+
+## Overview
+
+Panel Navigation is a feature for reading comics and manga that enables panel-by-panel navigation. Instead of viewing the entire page, readers can zoom into individual panels and navigate between them in a configurable reading order.
+
+This feature is particularly useful for:
+- **Comics/Manga**: Read panel-by-panel in the correct reading direction
+- **Small screens**: View panels at a comfortable zoom level on e-ink devices
+- **Detailed artwork**: Focus on individual panels without manual zooming
```
Comment (Frenzie) [inline doc/PanelNavigation.md]: I assume this document was written by AI?

File: doc/Unit_tests.md [hunk 2]
```diff
@@ -9,6 +9,16 @@ You can run them with `./kodev test`, examples:
...
 Check the output of `./kodev test -h` for the full usage.
+
+## Notable Test Files
+
+| Test File | Description |
+|-----------|-------------|
+| `readerpanelnav_spec.lua` | Panel navigation for comics/manga (sorting, merging, filtering) |
+| `readerhighlight_spec.lua` | Text selection and highlighting |
+| `readerpaging_spec.lua` | Page-based navigation |
+| `koptinterface_spec.lua` | K2pdfopt interface for document processing |
```
Comment (Frenzie) [inline doc/Unit_tests.md]: Hm, are they? I'd just leave this out.

File: doc/PanelNavigation.md (hunk 1 above)
Comment (vmanojkumar) [reply doc/PanelNavigation.md]: Of Course :)

Comment (Frenzie) [reply doc/PanelNavigation.md]: At the very least it's quite verbose. I'd drop most of it if not the entire document. Imo there's little use in duplicating what a cursory glance at the code also says, while the parts that aren't developer-centric seem fine but don't really belong in this location either. Pinging @poire-z

**Comment by poire-z** [comment]
>> I assume this document was written by AI?

> Of Course :)

Well, it's not obvious PRs are to be written by AI (even if I guess, when it looks clean, it must be written by AI....)
It would have been polite and useful to state these facts in the top post: how much were written by AI, how much you were and have become familiar with the codebase, and a few lines about the process you used to reach that point.
If we are to review the PR, it's good we know if we are addressing you to do the tweaks, or you are just being a proxy to that AI. And how much you will be able to continue adressing related issues.

Anyway (from my small experience), AI adds a lot of comments to state what they do or why some incremental change, which is good to know why they did that. In the end, the incremental comments may not have much context, so some of them can go.

Anyway, not much familiar with the existing panel zoom or reading comics to comment much on the feature. From a quick glance, it looks rather clean, the intrusion is minimal, parts read well separated (= it does not complicate the reading of the existing paths), there is a new module (that I won't even try to comprehend), things get triggered only from `ReaderHighlight:onPanelZoom()` so I guess the feature must be enabled and there are little risks of the new stuff imposing itself on normal PDF reading and its various complex modes (crop, reflow, etc...)
So, at least from a first glance, nothing against - but not familiar enough with Comics reading to do any approval.

**Comment by Frenzie** [comment]
Yes, the code looks fine. I have some nits about the AI adding unnecessary comments but that's neither here nor there. I was inquiring mostly about your opinion on that particular document. :-)

> (even if I guess, when it looks clean, it must be written by AI....)

Nah, it's more that it's extremely verbose about things that at best no one ever wants to see, at worst are misleading.

File: spec/unit/readerpanelnav_spec.lua [hunk 3]
```diff
@@ -0,0 +1,652 @@
+describe("ReaderPanelNav module", function()
+    local ReaderPanelNav
+
+    setup(function()
+        require("commonrequire")
+        -- We need to mock the UI dependencies since we're testing the module in isolation
+        ReaderPanelNav = require("apps/reader/modules/readerpanelnav")
```
Comment (Frenzie) [inline spec/unit/readerpanelnav_spec.lua:6]:
```suggestion
        require("commonrequire")
        ReaderPanelNav = require("apps/reader/modules/readerpanelnav")
```

File: frontend/apps/reader/modules/readerpanelnav.lua [hunk 4]
```diff
@@ -0,0 +1,1012 @@
...
+        LRBT = _("Left to right, bottom to top"),
+        RLTB = _("Right to left, top to bottom"),
+        RLBT = _("Right to left, bottom to top"),
+        TBLR = _("Top to bottom, left to right"),
+        TBRL = _("Top to bottom, right to left"),
+        BTLR = _("Bottom to top, left to right"),
+        BTRL = _("Bottom to top, right to left"),
+    }
+    return descriptions[direction] or direction
+end
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readerpanelnav.lua:140]: As a reference (mostly for me, but possibly not depending on how much manual vs AI was involved), the strings came from here. Presumably it's fine to simply duplicate them like this.

https://github.com/koreader/koreader/blob/ea6089b6d88284b6eda4f69e8b7407c15897c8ee/frontend/ui/data/koptoptions.lua#L307-L328

File: frontend/apps/reader/modules/readerpanelnav.lua [hunk 5]
```diff
@@ -0,0 +1,1012 @@
...
+ReaderPanelNav.onPhysicalKeyboardConnected = ReaderPanelNav.registerKeyEvents
+
+function ReaderPanelNav:onReaderReady()
+    self:setupTouchZones()
+end
+
+function ReaderPanelNav:setupTouchZones()
+    if not Device:isTouchDevice() then return end
+    -- Touch zones can be added here if needed for swipe-based panel navigation
+end
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readerpanelnav.lua]: Leftover.

```suggestion
```

File: frontend/apps/reader/modules/readerpanelnav.lua [hunk 6]
```diff
@@ -0,0 +1,1012 @@
...
+Panels that are completely outside the page are removed.
+
+@param panels array of panel rectangles {x, y, w, h}
+@param page_width width of the page
+@param page_height height of the page
+@treturn table array of clipped panels
+--]]
+function ReaderPanelNav:clipPanelsToPage(panels, page_width, page_height)
+    if not panels or #panels == 0 then
+        return panels
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readerpanelnav.lua]: Should it just `return` or `return false`? Otherwise you keep having to do that #panels == 0 check up the chain.

File: frontend/apps/reader/modules/readerpanelnav.lua [hunk 7]
```diff
@@ -0,0 +1,1012 @@
...
+
+    -- Return cached panels if available for current page
+    if self.current_page_panels and self.panels_page == pageno then
+        return self.current_page_panels
+    end
+
+    -- Check if document supports panel detection
+    if not self.ui.document.getAllPanelsFromPage then
+        return nil
+    end
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readerpanelnav.lua]: Assuming this is the right way to go about this, this should go first.

File: frontend/apps/reader/modules/readerpanelnav.lua [hunk 8]
```diff
@@ -0,0 +1,1012 @@
...
+--[[--
+Show a specific panel in the ImageViewer.
+
+@param panel panel rectangle {x, y, w, h}
+--]]
+function ReaderPanelNav:showPanel(panel)
+    if not panel then
+        logger.dbg("ReaderPanelNav: showPanel called with nil panel")
+        return
+    end
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readerpanelnav.lua]: Might as well make that a `dbg.dassert`. It feels like something that already should be filtered in the caller.

File: frontend/apps/reader/readerui.lua [hunk 9]
```diff
@@ -297,6 +298,13 @@ function ReaderUI:init()
             ui = self,
             document = self.document,
         })
+        -- panel navigation controller for comics/manga
+        self:registerModule("panelnav", ReaderPanelNav:new{
```
Comment (hius07) [inline frontend/apps/reader/readerui.lua:302]: Is there a way to avoid registering the module if a user doesn't read comics and would prefer to save some RAM?

For example, a plugin to zoom a panel
https://github.com/anezih/bubblezoom.koplugin/
Can be disabled or uninstalled if not used.

File: frontend/apps/reader/modules/readerhighlight.lua [hunk 10]
```diff
@@ -1657,6 +1734,106 @@ function ReaderHighlight:onPanelZoom(arg, ges)
...
+                if panelnav.current_panel_index == 0 then
+                    panelnav.current_panel_index = 1
+                end
+            end
+
+            -- Add keyboard shortcuts
+            if Device:hasKeys() then
+                imgviewer.key_events.NextPanel = { { "Right" }, event = "NextPanel" }
+                imgviewer.key_events.PrevPanel = { { "Left" }, event = "PrevPanel" }
+            end
```
Comment (Commodore64user) [inline frontend/apps/reader/modules/readerhighlight.lua]: I have an issue with just how haphazardly this is being added here, then again I won't be using this...

Comment (Frenzie) [reply frontend/apps/reader/modules/readerhighlight.lua]: Do you mean you'd add it on line 1740 for example or something else?

Comment (Commodore64user) [reply frontend/apps/reader/modules/readerhighlight.lua]: I mean there is already a function that amalgamates all key events in one place, but more importantly, this is simply doing the bare minimum for the emulator but does not really account for handling this on NT devices. 

on kindle the unit of progress is the page turners, this does nothing to fit into the existing way we do things, no RTL, it doesn't respect whether a user wants left/right to progress or not... as I said, the bare minimum

https://github.com/koreader/koreader/blob/ea6089b6d88284b6eda4f69e8b7407c15897c8ee/frontend/apps/reader/modules/readerpaging.lua#L53-L84

furthermore, does text selection work here? do these events clash with any existing ones? how does one exit this mode?

Comment (Frenzie) [reply frontend/apps/reader/modules/readerhighlight.lua]: Alright, so not so much haphazardly as simply incomplete. :-)

**Comment by Galunid** [comment]
I'd be very cautious with merging this PR, since it seems to use my panel detection algorithm. It had **a lot** of failure cases, but it was the best I could do given the hardware constraints. On top of my head it had issues with "diagonal" panels, small panels, overlapping panels, to name a few. This is generally acceptable since people don't zoom on every panel. For continuous reading however,  reliability of detection algorithm becomes very important. Missing panels/duplicating panels will be disruptive, at least in my opinion. This is probably a nice proof of concept, but I don't think it'll perform in the real world scenarios. 

I'm not really active here, so it's meant more as a heads up in case you're not particularly familiar with panel detection.

File: doc/PanelNavigation.md (hunk 1 above)
Comment (vmanojkumar) [reply doc/PanelNavigation.md]: I am removing it completely.. It can be easily generated again if needed.

File: doc/Unit_tests.md (hunk 2 above)
Comment (vmanojkumar) [reply doc/Unit_tests.md]: Removed.

File: spec/unit/readerpanelnav_spec.lua (hunk 3 above)
Comment (vmanojkumar) [reply spec/unit/readerpanelnav_spec.lua:6]: Removed.

File: frontend/apps/reader/modules/readerpanelnav.lua (hunk 4 above)
Comment (vmanojkumar) [reply frontend/apps/reader/modules/readerpanelnav.lua:140]: yeah, I did try to reuse it from the koptoptions.lua... but it was getting too much code. I have kept it as is for the time being.

File: frontend/apps/reader/modules/readerpanelnav.lua (hunk 5 above)
Comment (vmanojkumar) [reply frontend/apps/reader/modules/readerpanelnav.lua]: Removed

File: frontend/apps/reader/modules/readerpanelnav.lua (hunk 6 above)
Comment (vmanojkumar) [reply frontend/apps/reader/modules/readerpanelnav.lua]: Made it return nil

File: frontend/apps/reader/modules/readerpanelnav.lua (hunk 7 above)
Comment (vmanojkumar) [reply frontend/apps/reader/modules/readerpanelnav.lua]: Yes you are right. Fixed.

File: frontend/apps/reader/modules/readerpanelnav.lua (hunk 8 above)
Comment (vmanojkumar) [reply frontend/apps/reader/modules/readerpanelnav.lua]: Added.

File: frontend/apps/reader/readerui.lua (hunk 9 above)
Comment (vmanojkumar) [reply frontend/apps/reader/readerui.lua:302]: Let me think how to do it. The original panel stuff isn't part of any plugin and is always present. @Others please upvote if this is the way we want to go.

File: frontend/apps/reader/modules/readerhighlight.lua (hunk 10 above)
Comment (vmanojkumar) [reply frontend/apps/reader/modules/readerhighlight.lua]: Ah apologies... I had put everything I needed to do in the new file, just forgot to cleanup the caller. I have cleaned this up. Please check if it looks ok now.

**Comment by vmanojkumar** [comment]
> I'd be very cautious with merging this PR, since it seems to use my panel detection algorithm. It had **a lot** of failure cases, but it was the best I could do given the hardware constraints. On top of my head it had issues with "diagonal" panels, small panels, overlapping panels, to name a few. This is generally acceptable since people don't zoom on every panel. For continuous reading however, reliability of detection algorithm becomes very important. Missing panels/duplicating panels will be disruptive, at least in my opinion. This is probably a nice proof of concept, but I don't think it'll perform in the real world scenarios.
> 
> I'm not really active here, so it's meant more as a heads up in case you're not particularly familiar with panel detection.

Sorry, new to GitHub... the comment got closed by mistake. 
I agree that it doesnt work in many cases, I did tweak a few things and handled overlapping panel. I wrote this to see if it's any good. Seems to be working reasonably well for even badly scanned/old comics.

Would help if you can point some examples that are problematic. I can see if they work.

**Comment by Frenzie** [comment]
See https://github.com/koreader/koreader/issues/12487 for some discussion. From my perspective the problem is generally less detecting panels than the order they're in. But as far as merging this PR goes I don't think it's a problem if it only works reasonably well for some subset of comics.

File: frontend/apps/reader/readerui.lua (hunk 9 above)
Comment (poire-z) [reply frontend/apps/reader/readerui.lua:302]: If it stays a readermodule.lua, there's no reason to do differently than other readermodules and add complexity.
If it's a suggestion to make it a plugin, no real opinion.

File: frontend/apps/reader/modules/readerpanelnav.lua [hunk 11]
```diff
@@ -0,0 +1,1002 @@
...
+
+function ReaderPanelNav:registerKeyEvents()
+    if Device:hasKeys() then
+        -- P enters panel navigation view from the main view
+        self.key_events = {
+            EnterPanelNavMode = {
+                { "P" },
+                event = "EnterPanelNavMode",
+            },
+        }
```
Comment (Commodore64user) [inline frontend/apps/reader/modules/readerpanelnav.lua]: The whole point of hotkeys is that no shortcuts need to ever be hardcoded again. Not at the reader/file browser level at least.

Comment (vmanojkumar) [reply frontend/apps/reader/modules/readerpanelnav.lua]: Agree. Have moved the hotkeys, gesture handling within the readerpanelnav. Also followed the std structure similar to other views.

File: frontend/apps/reader/modules/readerpanelnav.lua [hunk 12]
```diff
@@ -0,0 +1,1251 @@
...
+
+    if image then
+        -- If we already have an ImageViewer open, just update the image
+        if self._panel_viewer then
+            logger.dbg("ReaderPanelNav: reusing existing ImageViewer")
+            -- Guard against updating while a repaint is in progress
+            if self._panel_viewer._updating then
+                logger.dbg("ReaderPanelNav: skipping update, already updating")
+                return
+            end
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readerpanelnav.lua:881]: Is this the correct behavior?

**Comment by vmanojkumar** [comment]
@Frenzie , I have made all changes I wanted to for this PR and the base PR. Can you please help merge this in ?

**Comment by Frenzie** [comment]
@Commodore64user Would you mind taking a look at the hotkey stuff again?

@vmanojkumar There are a few open comments.

**Comment by Commodore64user** [comment]
I'll have a look later, although there shouldn't be any... that was the point.

File: frontend/apps/reader/modules/readerhighlight.lua [hunk 13]
```diff
@@ -965,6 +952,83 @@ function ReaderHighlight:genPanelZoomMenu()
...
+                end,
+                hold_callback = function()
+                    G_reader_settings:saveSetting("panel_direction", code)
+                end,
+            })
+        end
+        -- Add panel navigation options at top level
+        table.insert(menu, {
+            text = _("Enable panel navigation"),
+            help_text = _("Navigate panel-by-panel through comics/manga using automatic panel detection. Use Left/Right arrow keys or tap left/right sides of zoomed panel to move between panels."),
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readerhighlight.lua:979]:
```suggestion
            help_text = _("Navigate through comics/manga panel-by-panel with automatic panel detection. Use the left/right arrow keys or tap the left/right sides of a zoomed panel to move between panels."),
```

File: frontend/apps/reader/modules/readerhighlight.lua [hunk 14]
```diff
@@ -965,6 +952,83 @@ function ReaderHighlight:genPanelZoomMenu()
...
+            callback = function()
+                panelnav.highlight_current_panel = not panelnav.highlight_current_panel
+                UIManager:setDirty(panelnav.dialog, "ui")
+            end,
+            enabled_func = function()
+                return self.panel_zoom_enabled and panelnav.panel_nav_enabled
+            end,
+        })
+        table.insert(menu, {
+            text = _("Show all detected panel boxes") .. " (" .. _("debug") .. ")",
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readerhighlight.lua:1018]: Just keep it as a single string. Being too clever just makes localization harder. :-)
```suggestion
            text = _("Show all detected panel boxes (debug)",
```

File: frontend/apps/reader/modules/readerhighlight.lua [hunk 15]
```diff
@@ -965,6 +952,83 @@ function ReaderHighlight:genPanelZoomMenu()
...
+            hold_callback = function()
+                G_reader_settings:saveSetting("panel_nav_enabled", not G_reader_settings:isTrue("panel_nav_enabled"))
+            end,
+            enabled_func = function()
+                return self.panel_zoom_enabled
+            end,
+        })
+        table.insert(menu, {
+            text_func = function()
+                return _("Panel reading direction") .. ": " .. panelnav.panel_direction
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readerhighlight.lua:995]: (better for localization)
```suggestion
                return T(_("Panel reading direction: %1"), panelnav.panel_direction)
```

File: frontend/apps/reader/modules/readerhighlight.lua [hunk 16]
```diff
@@ -965,6 +952,83 @@ function ReaderHighlight:genPanelZoomMenu()
...
     }
+    -- Add panel navigation options if the module is loaded
+    if self.ui.panelnav then
+        local panelnav = self.ui.panelnav
+        -- Panel reading direction submenu
+        local direction_codes = { "LRTB", "LRBT", "RLTB", "RLBT", "TBLR", "TBRL", "BTLR", "BTRL" }
+        local direction_submenu = {}
+        for _, code in ipairs(direction_codes) do
+            table.insert(direction_submenu, {
+                text = code .. " - " .. panelnav:getDirectionDescription(code),
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readerhighlight.lua:963]: (nicer typography)
```suggestion
                text = code .. " – " .. panelnav:getDirectionDescription(code),
```

File: frontend/apps/reader/modules/readerhighlight.lua [hunk 17]
```diff
@@ -965,6 +952,83 @@ function ReaderHighlight:genPanelZoomMenu()
...
+                panelnav.highlight_current_panel = not panelnav.highlight_current_panel
+                UIManager:setDirty(panelnav.dialog, "ui")
+            end,
+            enabled_func = function()
+                return self.panel_zoom_enabled and panelnav.panel_nav_enabled
+            end,
+        })
+        table.insert(menu, {
+            text = _("Show all detected panel boxes") .. " (" .. _("debug") .. ")",
+            help_text = _("Debug option: show green bounding boxes around all detected panels on the current page with panel numbers and coordinates. Current panel is shown in red."),
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readerhighlight.lua:1019]:
```suggestion
            help_text = _("Debug option: show green bounding boxes around all detected panels on the current page with panel numbers and coordinates. The current panel is shown in red."),
```

File: frontend/apps/reader/modules/readerpanelnav.lua [hunk 18]
```diff
@@ -0,0 +1,1266 @@
...
+                return true
+            end
+
+            -- Key bindings (replace default zoom with panel navigation)
+            if Device:hasKeys() then
+                imgviewer.key_events.ZoomIn = nil
+                imgviewer.key_events.ZoomOut = nil
+                imgviewer.key_events.NextPanel = { { { "RPgFwd", "LPgFwd", is_rtl and "Left" or "Right" } }, event = "NextPanel" }
+                imgviewer.key_events.PrevPanel = { { { "RPgBack", "LPgBack", is_rtl and "Right" or "Left" } }, event = "PrevPanel" }
+            end
```
Comment (Commodore64user) [inline frontend/apps/reader/modules/readerpanelnav.lua:1012]: absolutely not acceptable, this is still problematic, `hasFewKeys` not considered at all, I assume back is being overwritten which means, they are now trapped here forever. Also does not consider the left/right settings for dpads. Nope. no shortcuts which is good, but still doesn't cut it.

may I add, this whole `readerpanelnav` looks more like a third party plugin than a core file, stuff that should not be happening here is, like adding actions to dispatcher... I digress

Comment (Commodore64user) [reply frontend/apps/reader/modules/readerpanelnav.lua:1012]: had to double check the `hasFewKeys` and maybe is fine? I never know which button they use as back... anyway in reader rolling we do set all cursor keys to do something else so maybe this is fine... I'd still prefer to use the G_setting for using cursor keys on `hasDPad` and `useDPad`, actually I don't know what to think about these comic strips. What actions do you need here? do you not use a dictionary ever? are there any chapters or longer jumps than panels?

Comment (vmanojkumar) [reply frontend/apps/reader/modules/readerpanelnav.lua:1012]: I am getting your concern.  This being part of core and on top of ReaderUI is an issue. Even if I were to handle it now... there might be changes to handle some other new keys later on that will need to updated here. In fact one may argue that all std features must work as is in panelView => current implementation may not be appropriate. 
I can explore making an alternative way using a new option in ReaderZooming ( and call it panel ). 

Answer to your other questions : 
- What actions do you need here?  Zoomin, Zoomout, NextPanel, PrevPanel, Autorotate, Close ( the ImgViewer on tap opens up a close menu. But you are right, we might get stuck in this. But this would be no different from what happened earlier )

- do you not use a dictionary ever? In the normal readerUI it does seem long press on an image-area that has text underneath triggers OCR and folks might want to have it on panel view as well.

- are there any chapters or longer jumps than panels?  I don't think we should have it in panel view.


I could do the following if its ok.. this handles all that we see now, but might not in future
```

function ReaderPanelNav:registerKeyEvents()
    -- Check if user wants left/right for page turning instead
    if G_reader_settings:isTrue("left_right_keys_turn_pages") then
        -- Don't use left/right for panel nav
        -- Use up/down instead for devices with D-Pad
        if Device:hasDPad() and Device:useDPadAsActionKeys() then
            self.key_events.NextPanel = { { "Down" }, event = "NextPanel" }
            self.key_events.PrevPanel = { { "Up" }, event = "PrevPanel" }
        end
        return
    end
    -- User doesn't need left/right for pages, so we can use them for panels
    if Device:hasDPad() and Device:useDPadAsActionKeys() then
        -- Has D-Pad: use left/right with RTL awareness
        local nextKey = BD.mirroredUILayout() and "Left" or "Right"
        local prevKey = BD.mirroredUILayout() and "Right" or "Left"
        self.key_events.NextPanel = { { nextKey }, event = "NextPanel" }
        self.key_events.PrevPanel = { { prevKey }, event = "PrevPanel" }
    elseif Device:hasKeys() and not Device:hasFewKeys() then
        -- Has regular buttons, can use left/right
        local nextKey = BD.mirroredUILayout() and "Left" or "Right"
        local prevKey = BD.mirroredUILayout() and "Right" or "Left"
        self.key_events.NextPanel = { { nextKey }, event = "NextPanel" }
        self.key_events.PrevPanel = { { prevKey }, event = "PrevPanel" }
    elseif Device:hasFewKeys() then
        -- Very few buttons (like some older Kindles)
        -- Use up/down instead of left/right
        self.key_events.NextPanel = { { "Down" }, event = "NextPanel" }
        self.key_events.PrevPanel = { { "Up" }, event = "PrevPanel" }
    end
    -- Always provide an exit key (Back/Menu should always work)
    self.key_events.ClosePanelNav = { { "Back" }, event = "ClosePanelNav" }
end

```

Comment (Commodore64user) [reply frontend/apps/reader/modules/readerpanelnav.lua:1012]: Let me start by saying that I do not read comic books, so I am not entirely sure of a reader's specific needs. However, I have the following observations:

1. Button Availability: You seem to be acting under the assumption that all buttons are available for mapping, but this is not the case. Since this is becoming part of "core", it must work alongside other modules within the reader app. All buttons are already mapped. My assumption (correct me if I am wrong) is that this is a module similar to rolling or paging, where modes are mutually exclusive, which explains why they share events for page turning and the like. If this is not a mutually exclusive mode, there are no buttons available to give you.
2. Hardware Testing: I am not convinced you have tested this on an actual keyboard. Are you using the emulator? If so, run it with `./kodev run -t` and check the current button assignments.
3. User Flow: What exactly are the requirements here? As I asked earlier, once a user enters this mode, what happens next? How do they return to full-page mode? Again, check the emulator's behaviour.
4. Navigation Logic: The way I see it, you would likely need to use the page-turn buttons to move through panels and the left/right cursor keys for other actions. We need to test how the emulator currently behaves under your "panel view" first.
5. File Formats: What format are these comics in? Would this panel view be available for PDF, ePub, or something else?

Comment (Darthagnon) [reply frontend/apps/reader/modules/readerpanelnav.lua:1012]: > How do they return to full-page mode?

I tested https://github.com/JorgeTheFox/koreader-dynamic-panelzoom on my Kindle 4 (button-based) and it soft-locks the Kindle: I was able to enter panel zoom, but was not able navigate to next/previous panels, nor was I able to exit it.

https://github.com/JorgeTheFox/koreader-dynamic-panelzoom/issues/14
